examples/controlnet/train_controlnet_sd3.py

```diff
@@ -0,0 +1,1396 @@
+
+        def load_model_hook(models, input_dir):
+            while len(models) > 0:
+                # pop models so that they are not loaded again
+                model = models.pop()
+
+                # load diffusers style into model
+                load_model = SD3ControlNetModel.from_pretrained(input_dir, subfolder="controlnet")
+                model.register_to_config(**load_model.config)
+
+                model.load_state_dict(load_model.state_dict())
+                del load_model
+
+        accelerator.register_save_state_pre_hook(save_model_hook)
+        accelerator.register_load_state_pre_hook(load_model_hook)
+
+    if args.gradient_checkpointing:
+        controlnet.enable_gradient_checkpointing()
+
+    # Check that all trainable models are in full precision
+    low_precision_error_string = (
+        " Please make sure to always have all model weights in full float32 precision when starting training - even if"
+        " doing mixed precision training, copy of the weights should still be float32."
+    )
+
+    if unwrap_model(controlnet).dtype != torch.float32:
+        raise ValueError(
+            f"Controlnet loaded as datatype {unwrap_model(controlnet).dtype}. {low_precision_error_string}"
+        )
+
+    # Enable TF32 for faster training on Ampere GPUs,
+    # cf https://pytorch.org/docs/stable/notes/cuda.html#tensorfloat-32-tf32-on-ampere-devices
+    if args.allow_tf32:
+        torch.backends.cuda.matmul.allow_tf32 = True
+
+    if args.scale_lr:
+        args.learning_rate = (
+            args.learning_rate * args.gradient_accumulation_steps * args.train_batch_size * accelerator.num_processes
+        )
+
+    # Use 8-bit Adam for lower memory usage or to fine-tune the model in 16GB GPUs
+    if args.use_8bit_adam:
+        try:
+            import bitsandbytes as bnb
+        except ImportError:
+            raise ImportError(
+                "To use 8-bit Adam, please install the bitsandbytes library: `pip install bitsandbytes`."
+            )
+
+        optimizer_class = bnb.optim.AdamW8bit
+    else:
+        optimizer_class = torch.optim.AdamW
+
+    # Optimizer creation
+    params_to_optimize = controlnet.parameters()
+    optimizer = optimizer_class(
+        params_to_optimize,
+        lr=args.learning_rate,
+        betas=(args.adam_beta1, args.adam_beta2),
+        weight_decay=args.adam_weight_decay,
+        eps=args.adam_epsilon,
+    )
+
+    train_dataset = make_train_dataset(args, tokenizer_one, tokenizer_two, tokenizer_three, accelerator)
+
+    train_dataloader = torch.utils.data.DataLoader(
+        train_dataset,
+        shuffle=True,
+        collate_fn=collate_fn,
+        batch_size=args.train_batch_size,
+        num_workers=args.dataloader_num_workers,
+    )
+
+    tokenizers = [tokenizer_one, tokenizer_two, tokenizer_three]
+    text_encoders = [text_encoder_one, text_encoder_two, text_encoder_three]
+
+    def compute_text_embeddings(prompt, text_encoders, tokenizers):
```

Comment:
Would it be possible to update the script to use `compute_text_embeddings()` after the dataset is prepared? This way, we won't have to keep the text encoders in GPU memory during the course of training. 

Or, we could offload the text encoders to CPU (assuming enough RAM) after text embeddings are computed and then load them back as needed. 

WDYT?

Reply:
Yeah, you are right. I have updated the codes, and now we don't keep the text encoders in GPU memory.

Reply:
Hi, 
I have tried the script but I cannot remove text encoders out of GPU before training the controlnet, did you test this feature successfully? 
Thanks so much!

Reply:
Yes. In line 1133 of the code, text encoders and tokenizers have been deleted before training. https://github.com/huggingface/diffusers/blob/6dc6486565ea1d8d1be567eefc1094e9185560a1/examples/controlnet/train_controlnet_sd3.py#L1133C5-L1133C61

Reply:
It cannot work on my side, the text encoders still occupy ~10GB memory and I don't know why.
Another question is do you know where the text embedding (output of text encoders) saved ? In RAM or Disk?
Thanks so much for the reply!

Reply:
Are you sure it is the text encoders that occupy the memory? The GPU memory can be other models, as the vae, transformer, and controlnet models are still in memory. Besides, as we periodically run the validation, the text encoders will also be loaded every `validation_steps` steps. From my experiments, previously I need to separate the training and validation in two distinct GPUS, and after the above update I only need one GPU to run the script.

During training, the text embeddings from text encoders are in memory, though there is a cached one in disk such that it will not compute them in your next run as long as the configs are the same.